File: ctci/tmp/chapter_4_trees_graphs/c/4.4-check_balanced.c
#include "tree.h"
#include "binary_search_tree.h"
#include "limits.h"

/**
 * Calculates differences between heights of left and right nodes.
 * If difference is greater than 1, return INT_MIN.
 */
int check_height(tree_node *root)
{
    int left_height, right_height, difference;

    if (!root)
        return -1;
    left_height = check_height(root->left);
    if (left_height == INT_MIN)
        return INT_MIN;
    right_height = check_height(root->right);
    if (right_height == INT_MIN)
        return INT_MIN;
    difference = left_height - right_height;
    if (abs(difference) > 1)
        return INT_MIN;
    return MAX(left_height, right_height) + 1;
}

/**
 * Determines if a binary tree is balanced. 
 */
int check_balanced(tree_node *root)
{
    return check_height(root) != INT_MIN;
}

int main(void)
{
    tree_node *root = NULL;
    tree_node *n5, *n15, *n12;
    int arr[] = {20, 10, 30, 5, 15};
    int size = sizeof(arr) / sizeof(int), i;

    for (i = 0; i < size; i++)
        insert(&root, arr[i]);
    printf("%i\n", check_balanced(root)); // 1
    insert(&root, 17);
    printf("%i\n", check_balanced(root)); // 0

    return 0;
}

File: ctci/tmp/chapter_4_trees_graphs/js/4.4-checkBalanced.js
const { TreeNode } = require('./tree')

/**
 * Determines if a binary tree is balanced.
 * @param   {TreeNode}  root
 * @returns {boolean}
 */
const checkBalanced = root => {
  return checkHeight(root) != Number.MIN_VALUE
}

/* Calculates differences between heights of left and right nodes.
 * If difference is greater than 1, return INT_MIN */

/**
 * Calculates differences between heights of left and right nodes.
 * If difference is greater than 1, return INT_MIN.
 * @param   {TreeNode}  root
 * @returns {number}
 */
const checkHeight = root => {
  if (!root) return -1
  let leftHeight = checkHeight(root.left)
  if (leftHeight === Number.MIN_VALUE) return Number.MIN_VALUE
  let rightHeight = checkHeight(root.right)
  if (rightHeight === Number.MIN_VALUE) return Number.MIN_VALUE
  let difference = leftHeight - rightHeight
  if (Math.abs(difference) > 1) return Number.MIN_VALUE
  return Math.max(leftHeight, rightHeight) + 1
}

const n1 = new TreeNode(1)
const n2 = n1.addLeft(2)
const n3 = n1.addRight(3)
const n4 = n2.addLeft(4)
const n5 = n2.addRight(5)
const n6 = n3.addLeft(6)
const n8 = n4.addLeft(8)
const n9 = n4.addRight(9)
const n10 = n5.addLeft(10)
const n11 = n5.addRight(11)
const n12 = n6.addLeft(12)
const n13 = n6.addRight(13)

console.log(checkBalanced(n1)) // false
